java/debugger.jpda/src/org/netbeans/modules/debugger/jpda/models/ShortenedStrings.java

```diff
@@ -171,14 +212,43 @@ static String getStringWithLengthControl(StringReference sr) throws InternalExce
             } else {
                 assert sa != null;
                 int l = AbstractObjectVariable.MAX_STRING_LENGTH;
+                if (isCompressedImpl && isUTF16){
+                    l *= 2;
+                }
                 List<Value> values = ArrayReferenceWrapper.getValues(sa, 0, l);
                 char[] characters = new char[l + 3];
-                for (int i = 0; i < l; i++) {
-                    Value v = values.get(i);
-                    if (!(v instanceof CharValue)) {
-                        return "<Unreadable>";
+                if (isCompressedImpl) {
+                    //java compressed string
+                    if (!isUTF16) {
+                        //we can just cast to char
+                        for (int i = 0; i < l; i++) {
+                            Value v = values.get(i);
+                            if (!(v instanceof ByteValue)) {
+                                return ERROR_RESULT;
+                            }
+                            char c = (char)((ByteValue) v).byteValue();
+                            //remove the extended sign
+                            c &= 0xFF;
+                            characters[i] = c;
+                        }
+                    }
+                    else {
+                        //life is pain
+                        //We can't just inline code for this since... native
+                        //jazz and big/little endian stuff... so...
+                        //see StringUTF16.java
+                        //implement later!
+                        return "<Not Implemented>";
```

Comment:
To correctly decode a string we'd need to know the byte order of the target VM, but the following implemention should work correctly on little endian architectures (x86, arm, riscv), which should at least cover mainline onces.

```suggestion
                        // This assumes little endian encoding. This should work
                        // for most architectures (x86, arm, riscv), but will
                        // result in bogus data on big endian architectures
                        for (int i = 0; i < l; i++) {
                            int index = i * 2;
                            Value v = values.get(index);
                            if (!(v instanceof ByteValue)) {
                                return ERROR_RESULT;
                            }
                            Value v2 = values.get(index + 1);
                            if (!(v instanceof ByteValue)) {
                                return ERROR_RESULT;
                            }
                            char c1 = (char) ((ByteValue) v).byteValue();
                            char c2 = (char) ((ByteValue) v2).byteValue();
                            //remove the extended sign
                            c1 = (char) (0xFF & c1);
                            c2 = (char) (0xFF & c2);
                            // char bigEndianChar = (char) ((c1 << 8) | c2);
                            char litteEndianChar = (char) ((c2 << 8) | c1);
                            characters[i] = litteEndianChar;
                        }
```

Reply:
Should we try to check the high and low byte shifts to determine which one it is or not worry about it?

Reply:
I  thought about that, but the problem is, that there is no indicator, what is "the right" shift. In my testcase, both shifts resulted in a valid code point. You could ask the target VM to encode a string, but then you can also directly determine the byteorder there.